src/middleware/CompanyAuthMiddleware.ts

```diff
@@ -0,0 +1,107 @@
+
+import { Session, SessionStore } from 'ch-node-session-handler';
+import { SessionKey } from 'ch-node-session-handler/lib/session/keys/SessionKey';
+import { Cookie } from 'ch-node-session-handler/lib/session/model/Cookie';
+import { ISignInInfo } from 'ch-node-session-handler/lib/session/model/SessionInterfaces';
+import { NextFunction, Request, RequestHandler, Response } from 'express';
+import { inject} from 'inversify';
+import { BaseMiddleware } from 'inversify-express-utils';
+
+import { loggerInstance } from 'app/middleware/Logger';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { CompanyAuthConfig } from 'app/models/CompanyAuthConfig';
+import { Mutable } from 'app/models/Mutable';
+import { SessionStoreConfig } from 'app/models/SessionConfig';
+import { JwtEncryptionService } from 'app/modules/jwt-encryption-service/JwtEncryptionService';
+
+export class CompanyAuthMiddleware extends BaseMiddleware {
+
+    constructor(@inject(SessionStore) private readonly sessionStore: SessionStore,
+                private readonly encryptionService: JwtEncryptionService,
+                private readonly authConfig: CompanyAuthConfig,
+                private readonly sessionStoreConfig: SessionStoreConfig,
+                private readonly featureFlag: string) {
+        super();
+    }
+
+    public handler: RequestHandler = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+
+        if(this.featureFlag === '0'){
+            return next();
+        }
+
+        if (!req.session) {
+            return next(new Error('Session Expected but was undefined'));
+        }
+
+        const applicationData: ApplicationData = req.session
+            .getExtraData(APPLICATION_DATA_KEY) || {} as ApplicationData;
+
+        const companyNumber: string = applicationData.appeal.penaltyIdentifier.companyNumber;
+        const signInInfo: ISignInInfo | undefined = req.session.get<ISignInInfo>(SessionKey.SignInInfo);
+
+        if (this.isAuthorisedForCompany(signInInfo, companyNumber)){
+            loggerInstance().info(`CompanyAuthMiddleware: User is authenticated for ${companyNumber}`);
+            return next();
+        }
+
+        try {
+            const uri = await this.getAuthRedirectUri(req, res, companyNumber);
+
+            loggerInstance().debug(`CompanyAuthMiddleware: Redirecting to ${uri}`);
+            return res.redirect(uri);
+
+        } catch (err){
+            console.log(err);
+            next(err);
+        }
+    }
+
+    async getAuthRedirectUri(req: Request, res: Response, companyNumber: string): Promise<string> {
+
+        const originalUrl: string = req.originalUrl;
+        const scope: string = this.authConfig.oathScopePrefix + companyNumber;
```

Comment:
How is scope verified in the process? Does it matter? What would have happened if value provided by us was wrong?

Reply:
This documentation explains the role of scope and their values quite well: https://companieshouse.atlassian.net/wiki/spaces/Arch/pages/1334771713/API+Filing+Service+Designs+3rd+Parties+and+CH

Reply:
The convention for OAuth scopes is to provide the full URI of the resource including the domain then, add a period followed by the permission level e.g. 

the scope “https://api.companieshouse.gov.uk/user/profile.read” would provide the requesting client access to READ the user profile resource.

Reply:
What happens if you provide bad value, e.g. `foo bar` as a scope?

Reply:
The scope is the concatenation of  OATH_SCOPE_PREFIX and companyNumber
If you set the environment variable OATH_SCOPE_PREFIX to an incorrect value, it will break when the user is intercepted by the middleware